aiida_quantumespresso/calculations/pp.py

```diff
@@ -1,27 +1,218 @@
+
+    def prepare_for_submission(self, folder): # pylint: disable=too-many-branches,too-many-statements
+        """Create the input files from the input nodes passed to this instance of the `CalcJob`.
+
+        :param folder: an `aiida.common.folders.Folder` to temporarily write files on disk
+        :return: `aiida.common.datastructures.CalcInfo` instance
+        """
+
+        # Put the first-level keys as uppercase (i.e., namelist and card names) and the second-level keys as lowercase
+        if 'parameters' in self.inputs:
+            parameters = _uppercase_dict(self.inputs.parameters.get_dict(), dict_name='parameters')
+            parameters = {k: _lowercase_dict(v, dict_name=k) for k, v in six.iteritems(parameters)}
+        else:
+            parameters = {}
+
+        # Set default values. NOTE: this is different from PW/CP
+        for blocked in self._blocked_keywords:
+            namelist = blocked[0].upper()
+            key = blocked[1].lower()
+            value = blocked[2]
+
+            if namelist in parameters:
+                if key in parameters[namelist]:
+                    raise exceptions.InputValidationError(
+                        "You cannot specify explicitly the '{}' key in the '{}' "
+                        'namelist.'.format(key, namelist))
+            else:
+                parameters[namelist] = {}
+            parameters[namelist][key] = value
+
+        # Check for essential keys
+        try:
+            plot_num = parameters['INPUTPP']['plot_num']
+        except KeyError:
+            raise exceptions.InputValidationError("'[INPUTPP][plot_num]' must be explicitly set")
+
+        try:
+            dimension = parameters['PLOT']['iflag']
+        except KeyError:
+            raise exceptions.InputValidationError("'[PLOT][iflag]' must be explicitly set")
+
+        # Check that a valid plot type is requested
+        if plot_num in range(23) and plot_num not in [14, 15, 16]: # Must be integer in range 0-22, but not 14-16:
+            parameters['INPUTPP']['plot_num'] = int(plot_num) # If this test passes, we can safely cast to int
+        else:
+            raise exceptions.InputValidationError("'plot_num' must be an integer in the range 0-23")
+
+        # Check for valid plot dimension:
+        if dimension in range(5): # Must be in range 0-4:
+            parameters['PLOT']['iflag'] = int(dimension)
+        else:
+            raise exceptions.InputValidationError("'iflag' must be an integer in the range 0-4")
+
+        # Restrict the plot output to the file types that we want to be able to parse
+        dimension_to_output_format = {
+            0: 0, # Spherical integration -> Gnuplot, 1D
+            1: 0, # 1D -> Gnuplot, 1D
+            2: 7, # 2D -> Gnuplot, 2D
+            3: 6, # 3D -> Gaussian cube
+            4: 0, # Polar on a sphere -> # Gnuplot, 1D
+        }
+        parameters['PLOT']['output_format'] = dimension_to_output_format[dimension]
```

Comment:
Somewhat related to the data file discussion below: If we expect that people use the "raw" output file it would make sense to allow manually specifying the `output_format`, and just parse only the ones we understand.

That would be in the spirit of "allow everything the code itself can do", but I'd also be fine with keeping this as it is for now. If there is a need for different output formats, it's a relatively straightforward (and backwards-compatible) change.

Reply:
What should the parser do in the first instance if it doesn't understand the format?
Just warn the user?

Reply:
Yeah, I guess either a `report` or `warning` - level message.

But again, I'd also be fine with just discarding and not allowing explicit `output_format` for now.

Long-term it might actually be nicer to have tools to convert `ArrayData` to whatever output format is needed - it seems silly to couple the storage to the visualization program.

Reply:
I'd be inclined to leave this for the future. The idea of a warning or report is fine, but I think the argument against this is similar to what @sphuber says in his report - the user will only discover this after they've created the nodes in the DB and the calculation has effecively failed in that it did something unexpected.

Very supportive of an export module for common plotting/visualisation tools - this could be an `aiida-core` feature.